Add limayaml param settings to provisioning script environment

jandubois opened the pull request:
They will be prefixed with `PARAM_`, so `param.FOO` becomes `PARAM_FOO`.

This is useful because parameter substitution happens when a template is instantiated, so `[ "{{.Param.ROOTFUL}}" = true ]` becomes `[ "true" = true ]` in the `cloud-init-output.log`.

This mechanism also works better when the parameter contains quotes, which would break a simplistic `FOO="{{.Param.FOO}}"`.

Test sample:

```yaml
param:
  FOO: 'foo"bar'

provision:
- mode: system
  script: |
    echo "$PARAM_FOO"
```

Output:

```
LIMA 2024-08-30T09:13:11-07:00| Executing /mnt/lima-cidata/provision.system/00000000
foo"bar
LIMA 2024-08-30T09:13:11-07:00| Exiting with code 0
```

Example of real-life usage (from #2515) will be to replace

```diff
-     # Setting shell variable makes it easier to read cloud-init-output.log
-     readonly ROOTFUL="{{.Param.ROOTFUL}}"
-     if [ "$ROOTFUL" = true ]; then
+     if [ "$PARAM_ROOTFUL" = true ]; then
```

jandubois reviewed (commented):
Needs docs!

Also wondering if `probes` should have access too (they currently don't have access to `LIMA_CIDATA_*` variables, I think).

jandubois commented:
I'm not proud of the hack to make the `$PARAM_*` variables available to the probes, but I couldn't figure out a cleaner way as long as we support arbitrary interpreters via the hashbang line.

Anyways, I think the PR is ready for review now.

norio-nomura reviewed (changes requested):
This change looks like something I considered before implementing #2498 but ultimately abandoned because it seemed too complex for me. I find it impressive and at the same time feel that it was indeed too challenging for me.

I haven't finished reading the `probes` hack yet, but I'll share what I understand so far.

jandubois commented on pkg/cidata/cidata.TEMPLATE.d/boot.sh:
```diff
@@ -10,7 +10,9 @@ WARNING() {
-while read -r line; do export "$line"; done <"${LIMA_CIDATA_MNT}"/lima.env
+while read -r line; do [ -n "$line" ] && export "$line"; done <"${LIMA_CIDATA_MNT}"/lima.env
+# shellcheck disable=SC2163
+while read -r line; do [ -n "$line" ] && export "$line"; done <"${LIMA_CIDATA_MNT}"/param.env
```
> This works on `mode: dependency` and `mode: system`, but not on `mode: boot` and `mode: user`.

Thanks for catching that! I forgot about `boot`, but thought it would work for `user`.

> That should be documented or implemented.

I've implemented it for all modes, and added a test in the `test-misc.yaml` template to verify that they all work.

jandubois commented on pkg/limayaml/validate_test.go:
```diff
@@ -41,6 +41,7 @@ func TestValidateParamIsUsed(t *testing.T) {
 		`mounts: [{"location": "/tmp", mountPoint: "/tmp/{{ .Param.name }}"}]`,
 		`provision: [{"script": "echo {{ .Param.name }}"}]`,
+		`provision: [{"script": "echo $PARAM_name }}"}]`,
 		`probes: [{"script": "echo {{ .Param.name }}"}]`,
```
I've added it.

jandubois commented:
> This change looks like something I considered before implementing #2498

This change just build on top of #2498, which is still needed so we can use `{{.Param.Key}}` in all the places that are not provisioning scripts or probes.

> I haven't finished reading the `probes` hack yet, but I'll share what I understand so far.

I've moved the hack to a separate function and added rather verbose documentation about it. It feels kind of too long, but maybe it is useful.

norio-nomura commented on pkg/hostagent/requirements.go:
```diff
@@ -41,9 +41,52 @@ func (a *HostAgent) waitForRequirements(label string, requirements []requirement
+	}
+
+	// TODO we should have a symbolic constant for `/mnt/lima-cidata`
+	exportParam := `while read -r line; do [ -n "$line" ] && export "$line"; done<<EOF\n$(sudo cat /mnt/lima-cidata/param.env)\nEOF\n`
```
Both Alpine and Ubuntu worked with the following approach:
```suggestion
	exportParam := `env $(sudo cat /mnt/lima-cidata/param.env) `
```

jandubois replied:
This doesn't seem to work if any of the values contain spaces:

```console
$ cat xxx.env
FOO=foo'bar baz
BAR=foo"bar baz

$ while read -r line; do export "$line"; done<xxx.env

$ printenv FOO
foo'bar baz

$ env $(cat xxx.env) bash -c 'printenv FOO'
env: baz: No such file or directory
```

norio-nomura commented on pkg/limayaml/validate.go:
```diff
@@ -315,6 +318,21 @@ func Validate(y *LimaYAML, warn bool) error {
 		warnExperimental(y)
 	}
+
+	// Validate Param settings
```
It might be useful to include tests in `validate_test.go` for catching or not catching these errors.

jandubois replied:
I've added tests for these validations, and for the `probe.script` check above.

norio-nomura reviewed (commented):
The implementation and comments for `prefixExportParam()` made it very clear. Thanks to that, I can suggest a simpler command. 
Adding tests to `test-misc.yaml` is also very clear and good.

jandubois commented on examples/default.yaml:
```diff
@@ -237,6 +237,7 @@ containerd:
 #   playbook: playbook.yaml
 
 # Probe scripts to check readiness.
+# The scripts run in user mode. They must start with a '#!' line.
```
Why/where should that be noted? I don't think the end-user needs to care about it beyond the fact that a shebang line is required.

jandubois commented on pkg/cidata/cidata.TEMPLATE.d/user-data:
```diff
@@ -84,6 +84,12 @@ ca_certs:
+  # We need to embed the params.env as a here-doc because /mnt/lima-cidata is not yet mounted
+  while read -r line; do [ -n "$line" ] && export "$line"; done <<'EOF'
+    {{- range $key, $val := $.Param }}
+  PARAM_{{ $key }}={{ $val }}
```
There is the pesky issue of how to do the quoting if `$val` contains quotes. This will work with the current approach:

```
PARAM_FOO=foo"bar
PARAM_BAR=foo'bar
```

How would you generate the correct code via the template?

The other argument would be that the code is mostly the same as the code used elsewhere to process `lima.env` and `param.env`.

norio-nomura replied:
I feel that value should be escaped with `shellescape.Quote()` both when creating `param.env` and when expanding into `user-data`. Essentially, what the `while read -r...` loop is doing is escaping, right? In `.env` files as well, it's common to quote the value on the right-hand side of the `=`. 🤔

jandubois replied:
> I feel that value should be escaped with `shellescape.Quote()` both when creating `param.env` and when expanding into `user-data`.

I consider `lima.env` (and now `param.env`) to be similar to `/etc/environment`. They are not shell scripts and don't follow shell syntax; they just list key/value pairs.

Maybe we can change them to be shell scripts instead, but there is the risk that we may be overlooking some edge case for which it won't work. Why change code that seems to have been working correctly for us for a long time? What is the benefit to offset the risk?

norio-nomura replied:
> I consider `lima.env` (and now `param.env`) to be similar to `/etc/environment`

Got it, I understand with this perspective. Thank you!

norio-nomura commented on pkg/limayaml/validate_test.go:
```diff
@@ -53,7 +133,7 @@ func TestValidateParamIsUsed(t *testing.T) {
 	}
 
-	// use "{{if .Param.rootful \"true\"}}{{else}}{{end}}"" in provision, probe, copyToHost, and portForward
+	// use "{{if eq .Param.rootful \"true\"}}…{{else}}…{{end}}" in provision, probe, copyToHost, and portForward
```
🙏🏻

norio-nomura commented on pkg/limayaml/validate_test.go:
```diff
@@ -64,7 +144,7 @@ func TestValidateParamIsUsed(t *testing.T) {
 		`copyToHost: [{"guest": "/tmp", "host": "/tmp/{{if eq .Param.rootful \"true\"}}rootful{{else}}rootless{{end}}"}]`,
 		`portForwards: [{"guestSocket": "{{if eq .Param.rootful \"true\"}}/var/run{{else}}/run/user/{{.UID}}{{end}}/docker.sock", "hostSocket": "{{.Dir}}/sock/docker.sock"}]`,
-		`portForwards: [{"guestSocket": "/var/run/docker.sock", "hostSocket": "{{.Dir}}/sock/docker-{{if eq .Param.rootful \"true\"}}rootfule{{else}}rootless{{end}}.sock"}]`,
+		`portForwards: [{"guestSocket": "/var/run/docker.sock", "hostSocket": "{{.Dir}}/sock/docker-{{if eq .Param.rootful \"true\"}}rootful{{else}}rootless{{end}}.sock"}]`,
```
🙏🏻

norio-nomura reviewed (approved):
LGTM!

jandubois commented:
> LGTM!

🙏 Thanks! All checks have passed, so I will merge now.

nirs commented:
This is awesome improvement, I wanted to ask how to inject values from the instance yaml to the cloud-init scripts.

However the names makes this harder to understand. Maybe rename to:

```yaml
env:
  MY_KEY: VALUE
provision:
  - mode: system
    script: |
      #!/bin/bash
      echo "MY_KEY"
```

No magic prefix, what you put in env added to the provision scripts.

This is also similar to docker/podman --env flag.

norio-nomura commented:
> Maybe rename to:

The `.env` field already exists. However, it had too broad of an impact to use solely for modifying the behavior of scripts within limayaml. That's why I wanted variables that could only be used in Lima, and after some consideration, I implemented `.param`.

nirs commented:
Yes being able to limit the scope is nice. Did you consider something like this?

```yaml
env:
  KEY: global
bootEnv:
  BOOT_KEY: boot
provision:
  - mode: system
    script: |
      echo $KEY       # -> gloabl
      echo $BOOT_KEY  # -> boot
probes:
  - mode: system
    script: |
      echo $KEY       # -> gloabl
      echo $BOOT_KEY  # -> boot
```

Since all provision steps are run by the same global boot.sh script, we can pass the bootEnv to the scripts by exporting the boot env vars from the script.

I did not check how probes are run, but since they are scripts there must be some shell running them and we can modify its environment.

etho201 commented:
Is it possible to define the parameters from an external file (for example, `variables.yaml`) in order to keep custom/sensitive parameters out of the template files?

jandubois commented:
> Is it possible to define the parameters from an external file (for example, `variables.yaml`) in order to keep custom/sensitive parameters out of the template files?

Not right now, but it will be once the implementation of the `basedOn` mechanism I described in https://github.com/lima-vm/lima/discussions/2520#discussioncomment-10193570 is complete. I'm working on it right now; the tricky part is merging YAML files while retaining comments and not just values.

You will be able to write your template like this:

```yaml
basedOn:
- ~/.variables.yaml
- template://fedora

provision:
…
```

Or you could have your private template that contains just the variables, and then reference a public base template:

```yaml
basedOn: [https://example.com/my-shared-template.yaml]

param:
  secret: password123
```

The settings from the base templates will be merged with your instance template when the instance is created, similar to `$LIMA_HOME/_config/default.yaml` is providing defaults when the instance starts.

That means values from the base templates act as defaults, but the settings in the main template take precedence.

It also means that the combined template will include your secrets, but it is stored in the Lima instance directory, which is only readable by your own user (permission 600).

nirs commented:
> Or you could have your private template that contains just the variables, and then reference a public base template:
> 
> ```yaml
> basedOn: [https://example.com/my-shared-template.yaml]
> 
> param:
>   secret: password123
> ```

This is not a good way to store secrets. We can add keychain integration so the owner of the secret can unlock the keychain or lima for accessing the secret when stating a cluster.

I looked at this for this issue:
https://github.com/nirs/oc-clusterset/issues/1

jandubois commented:
> This is not a good way to store secrets.

I agree, but @etho201 asked about secrets that are already in a text file. It is similar to `~/.aws` containing credentials.

> We can add keychain integration so the owner of the secret can unlock the keychain or lima for accessing the secret when stating a cluster.

As long as the secret eventually ends up in `~/.lima/INSTANCE/lima.yaml` it does not make a big difference; you still end up with the clear text on the disk. Even if not there, it will still be in the cloud-init data somehow, with the instance being able to access it.

So ideally you want something like ssh agent forwarding, but it all depends on how those secrets are going to be used during provisioning. I can't think of a generic solution.

jandubois commented:
BTW, if you want to continue this discussion, please move it to the [Discussions](https://github.com/lima-vm/lima/discussions). Having it at the bottom of the thread of an already closed pull request means it will not be very visible.
